apps/src/lib/client/tendermint_rpc_types.rs

```diff
@@ -40,77 +40,59 @@ impl TxResponse {
     /// Searches for custom events emitted from the ledger and converts
     /// them back to thin wrapper around a hashmap for further parsing.
     pub fn parse(
-        json: serde_json::Value,
+        event: Event,
         event_type: NamadaEventType,
         tx_hash: &str,
     ) -> Self {
-        let tx_hash_json = serde_json::Value::String(tx_hash.to_string());
-        let mut selector = jsonpath::selector(&json);
-        let mut index = 0;
+        let events = event
+            .events
+            .expect("We should have obtained Tx events from the RPC");
         let evt_key = event_type.to_string();
         // Find the tx with a matching hash
-        let hash = loop {
-            if let Ok(hash) =
-                selector(&format!("$.events.['{}.hash'][{}]", evt_key, index))
-            {
-                let hash = hash[0].clone();
-                if hash == tx_hash_json {
-                    break hash;
-                } else {
-                    index += 1;
+        macro_rules! tx_error {
+            () => {
+                || {
+                    eprintln!(
+                        "Couldn't find tx with hash {tx_hash} in events \
+                         {events:?}",
+                    );
+                    safe_exit(1)
                 }
-            } else {
+            };
+        }
+        let (index, _) = events
+            .get(&format!("{evt_key}.hash"))
+            .unwrap_or_else(tx_error!())
+            .iter()
+            .enumerate()
+            .find(|(_, hash)| hash == &tx_hash)
+            .unwrap_or_else(tx_error!());
+        let info = events[&format!("{evt_key}.info")][index].clone();
+        let log = events[&format!("{evt_key}.log")][index].clone();
+        let height = events[&format!("{evt_key}.height")][index].clone();
+        let code = events[&format!("{evt_key}.code")][index].clone();
+        let gas_used = events[&format!("{evt_key}.gas_used")][index].clone();
+        let initialized_accounts = events
+            [&format!("{evt_key}.initialized_accounts")]
+            .get(index)
+            .as_ref()
+            .map(|initialized_accounts| {
+                serde_json::from_str(initialized_accounts).unwrap()
+            })
+            .unwrap_or_else(|| {
                 eprintln!(
-                    "Couldn't find tx with hash {} in the event string {}",
-                    tx_hash, json
+                    "Tendermint omitted one of the expected indices in events"
```

Comment:
This is not an error. Most events should not initialize any accounts

Reply:
```
    "events": {
      "applied.code": [
        "0",
        "0",
        "1"
      ],
      "applied.initialized_accounts": [
        "[]",
        "[]"
      ],
      "tm.event": [
        "NewBlock"
      ],
      "applied.info": [
        "Transaction is valid. Gas used: 0; VPs result: ",
        "Transaction is valid. Gas used: 0; VPs result: ",
        "Tx rejected: Unsupported protocol transaction type"
      ],
      "applied.hash": [
        "3CEDD1EB1A23D3235B10F114F997EB05C3748F59DC69A62984DE8B0019166CEA",
        "B1D3BF49B333C4B80680301E7F5F4A7718B9E44BE74CD95A1E292FD2633E3EFB",
        "7793D575C0903458DD94BE8BAF0D79D19551A86FD38E9AA0152A8781E5E1DF47"
      ],
      "applied.log": [
        "",
        "",
        ""
      ],
      "applied.height": [
        "2",
        "2",
        "2"
      ],
      "applied.gas_used": [
        "0",
        "0",
        "0"
      ]
    }
```
if a tx is rejected (in this example, a valset upd protocol tx), tendermint omits one of the initialized accounts arrays from the RPC response. it's an error because we're relying on fixed indexes to join related data together, which is why we do that index lookup earlier in the file.

Reply:
a later pr should honestly just change the way we lookup event data. there's an alternative, more reliable way to do this, through the `EndBlock` data.

Reply:
https://p.sicp.me/k64uK example json data

Reply:
Oh, ok